### DISCUSSION TITLE
Setting path for generated content/originals in new installation

### DISCUSSION THREAD

**Discussion by aleyoscar** [immich-app/immich#2328] [Q&A] [answered]
Hello,

I see that in version 1.52, there is now the ability to separate the library from the cached content: "Users can now store thumbnails and encoded videos at a different location," but I am unsure how to set this up with a new server installation. When downloading the .env file as said in the installation documentation, I don't see a variable for the library/originals and the generated content/thumbnails, I only see the UPLOAD_LOCATION variable which I believe would house both the generated content and the library? How would I go about separating the two? For example I would like originals to be located at /mnt/files/photos, and any generated content at /mnt/generated.

Thanks!

**Comment by Allram** [comment, marked as answer]
This is how i have set up my .env to keep the thumbnails and other small files on my NVME cache drives and the rest on spinning disks:

```
###################################################################################
# Upload File Location
#
# This is the location where uploaded files are stored.
###################################################################################

LIBRARY_LOCATION=/mnt/user/Immich/
THUMBS_LOCATION=/mnt/apps/Immich-Cache/thumbs/
UPLOAD_LOCATION=/mnt/apps/Immich-Cache/upload/
PROFILE_LOCATION=/mnt/apps/Immich-Cache/profile/
VIDEO_LOCATION=/mnt/apps/Immich-Cache/encoded-video/
```


And then i added this volumes to the following dockers in docker-compose:
- immich-machine-learning
- immich-microservices
- immich-server

```
    volumes:
      - ${LIBRARY_LOCATION}:/usr/src/app/upload/library
      - ${UPLOAD_LOCATION}:/usr/src/app/upload/upload
      - ${THUMBS_LOCATION}:/usr/src/app/upload/thumbs
      - ${PROFILE_LOCATION}:/usr/src/app/upload/profile
      - ${VIDEO_LOCATION}:/usr/src/app/upload/encoded-video
   ```

**Reply by aleyoscar** [reply]
Works! Thanks

**Reply by dhruvinsh** [reply]
What does LIBRARY_LOCATION and UPLOAD_LOCATION represents? exiting media vs upload media (via app or some other means)?

**Reply by jrasm91** [reply]
All uploaded media hora you upload/ first (temporarily) then get moved into their final place in the library/ folder.

**Reply by patrontheo** [reply]
Thanks! Is it necessary to add those volumes to immich-machine-learning as well ? I get the sense that immich-microservices and immich-server is enough, no ?
EDIT: it is not.

**Reply by mcjak97** [reply]
Hello @Allram 
From what i understand, all this locations must be in external NAS share, yes? Or only LIBRARY_LOCATION?
LIBRARY_LOCATION=/mnt/user/Immich/
THUMBS_LOCATION=/mnt/apps/Immich-Cache/thumbs/
UPLOAD_LOCATION=/mnt/apps/Immich-Cache/upload/
PROFILE_LOCATION=/mnt/apps/Immich-Cache/profile/
VIDEO_LOCATION=/mnt/apps/Immich-Cache/encoded-video/
And what if i have 3 users, how can i separate external LIBRARY_LOCATION for every user?

**Reply by theOryol** [reply]
Adding this to help others like myself who come across this great post:

The final piece to the puzzle is to turn on the storage template feature in the Immich settings - (the front end settings, not the server backend). Go to Administration --> Settings --> Storage template --> Enabled (turn on)
Then go to "Jobs" and run "STORAGE TEMPLATE MIGRATION" to apply it to the existing files in the upload folder.

Once I did this, all the files made their way to the new library folder location.

**Reply by vzvl** [reply]
Tried it and for me it now creates a docker volume for /usr/src/app/upload which is kind of annoying as it gets recreated every time I down and up the container (I mean it should be empty, but still...). Any recommendations how to avoid the temporary creation of this unnecessary (empty) volume?

**Comment by Cwy0116** [comment]
The pth of the immich C:\Users\nasho.docker\immich-app
I want change to  X:\immich-app
So how do I transfer the path? In .env and other files

**Comment by Electron707** [comment]
Is this solution applicable to the latest immich (v2.x)? as in the official documentaiton i could not find any refrence to the LIBRARY_LOCATION....i tried to set it up but its not working as well...so not sure! any one still using this with the latest version. Ideally want the same solution as has been provided up...the actual Photo library resides in a mounted spinning HDD ... while the rest of the content remains seperate on an SSD. The only variable i can see is UPLOAD_LOCATION and then all items remain there.

Even with the EXTERNAL_LIBRARY function any uploaded content goes to the UPLOAD_LOCATION.
